admin/server/github.go

```diff
@@ -88,53 +134,254 @@ func (s *Server) githubConnect(w http.ResponseWriter, r *http.Request, pathParam
 	qry := r.URL.Query()
 	setupAction := qry.Get("setup_action")
-	if setupAction != "install" && setupAction != "update" { // TODO: Also handle "request"
+	if setupAction != "install" && setupAction != "update" && setupAction != "request" {
 		http.Error(w, fmt.Sprintf("unexpected setup_action=%q", setupAction), http.StatusBadRequest)
 		return
 	}
-	installationIDStr := qry.Get("installation_id")
-	installationID, err := strconv.Atoi(installationIDStr)
-	if err != nil || installationID == 0 {
-		http.Error(w, fmt.Sprintf("unexpected installation_id=%q", installationIDStr), http.StatusBadRequest)
-		return
-	}
 
-	// Check there's an authenticated user (this should always be the case for flows initiated by githubConnect)
 	claims := auth.GetClaims(r.Context())
 	if claims.OwnerType() != auth.OwnerTypeUser {
 		http.Error(w, "only authenticated users can connect to github", http.StatusUnauthorized)
 		return
 	}
 
-	// Associate the user with the installation
-	err = s.admin.ProcessUserGithubInstallation(r.Context(), claims.OwnerID(), int64(installationID))
+	remoteURL := qry.Get("state")
+	account, repo, ok := gitutil.SplitGithubURL(remoteURL)
+	if !ok {
+		// request without state can come in multiple ways like
+		// 	- if user changes app installation directly on the settings page
+		//  - if admin user accepts the installation request
+		// Nothing to be done in such cases. May be redirect to some rill page ??
+		w.WriteHeader(http.StatusOK)
+		return
+	}
+
+	if setupAction == "request" {
+		// access requested
+		redirectURL, err := url.Parse(s.urls.githubConnectRequest)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("failed to create connect request url: %s", err.Error()), http.StatusInternalServerError)
+			return
+		}
+
+		qry := redirectURL.Query()
+		qry.Set("remote", remoteURL)
+		redirectURL.RawQuery = qry.Encode()
+		http.Redirect(w, r, redirectURL.String(), http.StatusTemporaryRedirect)
+		return
+	}
+
+	// install/update setupAction
+
+	code := qry.Get("code")
+	if code == "" {
+		http.Error(w, "unauthorised user", http.StatusUnauthorized)
+		return
+	}
+
+	// exchange code to get an auth token and create a github client with user auth
+	githubClient, err := s.userAuthGithubClient(ctx, code)
+	if err != nil {
+		http.Error(w, "unauthorised user", http.StatusUnauthorized)
+		return
+	}
+
+	// verify there is no spoofing and the user is a collaborator to the repo
+	// There may not be a need for this because we are not storing any state wrt installation
+	// Keeping this for future proofness if we start storing some states
+	gitUser, isCollaborator, err := s.isCollaborator(ctx, account, repo, githubClient)
 	if err != nil {
-		http.Error(w, fmt.Sprintf("failed to track github install: %s", err), http.StatusInternalServerError)
+		http.Error(w, fmt.Sprintf("failed to verify ownership: %s", err), http.StatusUnauthorized)
 		return
 	}
 
+	if !isCollaborator {
+		http.Error(w, "unauthorised user", http.StatusUnauthorized)
+		return
+	}
```

Comment:
If the user doesn't have access to the remote, but granted access to some other repos, I think this case will be hit? If yes, it should be handled nicely by redirecting to the retry page (maybe they're logged into the wrong Github account).

Reply:
The users can't really grant access to a repo that they are not collaborator of so a user should not reach here under normal circumstances. This is only for cases when someone is trying to hit the API manually and do some spoofing.

Reply:
Are you sure? I think if a user has two Github accounts, they might clone locally using another than they're logged in with (by mistake, if they switch often).

Reply:
But they will only be able to install from the account where they have access to do so and the callback will be triggered with that account's code only so no chance of landing on the installation callback by another account.